src/layers/messages/opt-in/index.tsx

```
@@ -0,0 +1,58 @@
+import React, { useEffect } from 'react'
+import { useEntryLayerContext } from 'layers/entry'
+import { useAnalyticsLayerContext } from 'layers/analytics'
+import { useProductLayerContext } from 'layers/product'
+import { useCommsLayerContext } from 'layers/comms'
+import { useMessageLayerContext } from 'layers/messages/provider'
+
+
+type Children = { children: React.ReactNode }
+const OptInMessages = ({ children }: Children) => {
+  const { handleOptIn, onOptIn } = useEntryLayerContext()
+  const { report } = useAnalyticsLayerContext()
+  const { widgetURL } = useProductLayerContext()
+  const { registerMessageHandler, deregisterMessageHandler } = useCommsLayerContext()
+  const { widgetRef } = useMessageLayerContext()
+
+  useEffect(
+    () => {
+      registerMessageHandler({ origin: widgetURL, message: 'opt-in' }, async (data: any) => {
+        try {
+          const details = handleOptIn
+            ? (await handleOptIn())
+            : (onOptIn && data.externalId)
+              ? (await onOptIn(data.externalId))
+              : null
+
+          if (details) {
+            widgetRef!.current!.contentWindow!.postMessage(
```

should we maybe check for `widgetRef.current` and then if it doesn't exist log an unexpected ref error or something? instead of the series of `!`

---

oh hmm I intended this as something that would just get caught in the `try-catch`

it's kinda shorthand for actually doing the checks, bc it definitely seems bad to just do a silent failure if it doesn't exist (i.e., either do manual validation by checking each one exists, and throw an error if not, or just let it throw and catch it)

---

missed that this was in a try catch